### ISSUE TITLE
Invalid curve attacks

### ISSUE THREAD

**Issue by cbartle891** [tlswg/draft-deprecate-obsolete-kex#4]
What do you think about adding the detail about double-and-add algorithms to the explanation about invalid curve attacks to address this issue?: https://github.com/cbartle891/draft-deprecate-ffdh/issues/13

**Comment by nimia** [comment]
I'm actually not sure if use of the double-and-add algorithm is an essential requirement for invalid curve attacks.
I'm not an expert, but my understanding is:
1. The ECC code should verify that the point is on the curve.
2. The ECC code then computes a multiple of the point, through the double-and-add algorithm or through other algorithms.
I think absent step 1, attacks are generally possible, regardless of the algorithm used in step 2. (And reading the paper, the algorithm is only refered to in 4 places).
I can ask the authors of the paper about this - should I?

**Comment by cbartle891** [comment]
I'm not an expert in this either, but I think it depends on the reason we're saying that we want to deprecate ECDH. If it's because of side-channels, then we should mention the double-and-add algorithm, because that seems to be the thing that allows for side-channel attacks. If, however, we want to deprecate ECDH because implementations tend not to have the check you mention in step 1 (which does not appear to require a side-channel attack), then we don't need to mention the double-and-add algorithm.

But if we're only arguing the latter (that ECDH is weak because some implementations don't include the necessary check), that doesn't really seem like a strong enough reason if mpg is correct that the check is simple and already mandated by the standard.

Are there other ECC algorithms that aren't susceptible to side-channel attacks? If so, then I wonder if the case against ECDH because of side-channel attacks is rather weak since someone can just say "well, I don't use that algorithm." You could argue that this is a very subtle issue, and some people may use the double-and-add algorithm without realizing that that leaves them vulnerable to side-channel attacks, so we should just get rid of ECDH completely. But since ECDH is already not recommended because it doesn't provide forward secrecy, people are really already using it at their own risk.

**Comment by nimia** [comment]
The argument I had in mind is: In addition to not providing forward secrecy, using long-lived keys opens the door to a broad class of attacks, where the attacker finds a way to repeatedly query a secret value. This includes invalid curve attacks, as well as Filippo's "Carry Bit" attack:
https://www.blackhat.com/us-18/briefings/schedule/#squeezing-a-key-through-a-carry-bit-11773
These attacks are usually implementation dependent; invalid curve attacks and the Carry Bit Attack are. But broadly, we have a bad history with such attacks, not only for ECDH but generally (e.g. Raccoon).

Going with only the forward secrecy argument is a reasonable choice. But I think the term forward secrecy is usually interpreted as "what happens when someone steals the key". There's a more subtle issue of "is there a long-lived key that someone can hope to slowly extract through an attack".

So in conclusion, maybe we should describe and then refer to this category of attacks. rather than specifically invalid curve attacks? What do you think?

**Comment by cbartle891** [comment]
> maybe we should describe and then refer to this category of attacks. rather than specifically invalid curve attacks? What do you think?

Yeah, that sounds like the way to go. Enumerating all the issues with ECDH would be good, to build a strong case that creating a secure implementation of ECDH is rife with pitfalls and thus should be avoided. Is this all of them?:

1. Invalid curve attacks (due to not checking that the point is on the curve)
2. Side-channel attacks (due to use of the double-and-add algorithm)
3. The Carry Bit attack

Are there any others?

**Comment by nimia** [comment]
> Yeah, that sounds like the way to go

SGTM :-)

> Enumerating all the issues with ECDH would be good
> Are there any others?

Your list covers most attacks to my knowledge. Though I propose enumerating this slightly differently, since there were several distinct side-channel attacks throughout the years. 
I could do a literature review, enumerate specific attacks, and build a case against this category of attacks (where one could avoid the issue entirely if one doesn't use a long-lived key).
However, I can only really start the week of Dec 13th, so I'll push and ping you when there's draft text ready for review?

**Comment by cbartle891** [comment]
Sounds great!

**Comment by nimia** [comment]
I pushed a proposed draft of the new writeup.
I only included a single example of a side channel attack. Some implementers may consider these out of scope entirely, and some attacks are quite niche even if side channels are generally in scope, so I eventually thought we shouldn't harp on this too much.
What do you think?

Also, I hope you won't be reading this for at least a few days :-) Have a great new year's eve!

**Comment by cbartle891** [comment]
Thanks! XD Cool, I'll look it over.

**Comment by cbartle891** [comment]
In general, looks good to me! Posted a couple minor comments. Thanks for writing that up!
